pages/index.js

```diff
@@ -57,20 +57,164 @@ const ItemCard = styled.div`
   }
 `;
 
+const FormContainer = styled.form`
+  margin-bottom: 20px;
+  padding: 10px;
+  border: 1px solid #ddd;
+  border-radius: 5px;
+
+  label {
+    display: block;
+    margin-bottom: 5px;
+  }
+
+  input,
+  select {
+    width: 100%;
+    padding: 8px;
+    margin-bottom: 10px;
+    border-radius: 5px;
+    border: 1px solid #ddd;
+  }
+
+  button {
+    padding: 10px 20px;
+    background-color: #007bff;
+    color: white;
+    border: none;
+    border-radius: 5px;
+    cursor: pointer;
+
+    &:hover {
+      background-color: #0056b3;
+    }
+  }
+
+  .error {
+    color: red;
+    margin-bottom: 10px;
+  }
+`;
+
 const ShoppingList = () => {
-  const [items] = useState(
-    initialItems.map((item) => ({
-      ...item,
-      category: item.category.toLowerCase(),
-    }))
-  );
+  const [items, setItems] = useState([]);
   const [categories] = useState(initialCategories);
+  const [newItem, setNewItem] = useState({
+    name: "",
+    quantity: 1,
+    category: "",
+    comment: "",
+  });
+  const [errors, setErrors] = useState({});
+
+  useEffect(() => {
+    const storedItems = localStorage.getItem("shoppingItems");
+    if (storedItems) {
+      setItems(JSON.parse(storedItems));
+    } else {
+      setItems(
+        initialItems.map((item) => ({
+          ...item,
+          category: item.category.toLowerCase(),
+        }))
+      );
+    }
+  }, []);
+
+  useEffect(() => {
+    if (items.length > 0) {
+      localStorage.setItem("shoppingItems", JSON.stringify(items));
+    }
+  }, [items]);
+
+  const handleInputChange = (e) => {
+    const { name, value } = e.target;
+    setNewItem({ ...newItem, [name]: value });
+  };
+
+  const validateForm = () => {
+    const errors = {};
+    if (!newItem.name) errors.name = "Name is required";
+    if (!newItem.quantity || newItem.quantity <= 0)
+      errors.quantity = "Quantity must be a positive number";
+    if (!newItem.category || newItem.category === "default")
+      errors.category = "Please select a category";
+    return errors;
+  };
+
+  const handleSubmit = (e) => {
+    e.preventDefault();
+    const formErrors = validateForm();
+    if (Object.keys(formErrors).length > 0) {
```

Comment:
Since categories is  a static array like `["Dairy", "Bakery"]`, there is no need to use Object.keys(categories). You can directly map over the array, as this will be more efficient and clearer.